service/s3/s3crypto/s3crypto_client.go

```diff
@@ -0,0 +1,175 @@
+package s3crypto
+
+import (
+	"encoding/hex"
+	"io/ioutil"
+	"log"
+	"os"
+
+	"github.com/aws/aws-sdk-go/aws/request"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+)
+
+// Client is an S3 crypto client. By default we will use EncryptionOnly mode which
+// will use AES Wrap for key wrapping and AES GCM for content encryption.
+// AES GCM will load all data into memory. However, the rest of the content algorithms
+// do not load the entire contents into memory.
+type Client struct {
+	S3     *s3.S3
+	Config Config
+}
+
+// Config used to customize the Client
+type Config struct {
+	// Mode is used to dictate how we encrypt and decrypt the data.
+	//
+	// Defaults to AES GCM with the suggested pairing of AES Wrap/PRSA/KMS
+	Mode CryptoMode
+	// SaveStrategy will dictate where the envelope is saved.
+	//
+	// Defaults to the object's metadata
+	SaveStrategy
+	// InstructionSuffix is the instruction file name suffix. If it is empty, then
+	// the item key will be used followed by .instruction
+	InstructionSuffix string
+	// This is used to instantiate new kms clients when calling GetObject
+	KMSSession *session.Session
+	S3Session  *session.Session
+
+	// Used for instantiating non-kms key providers when getting objects
+	MasterKey []byte
+}
+
+// New instantiates a new S3 crypto client
+// Example:
+// cmkID := "some key id to kms"
+// kp, err = s3crypto.NewKMSKeyProvider(session.New(), cmkID, s3crypto.NewJSONMatDesc())
+// if err != nil {
+//   return err
+// }
+// svc := s3crypto.New(s3crypto.Authentication(kp))
+func New(mode CryptoMode, options ...func(*Client)) *Client {
+	sess := session.New()
```

Comment:
New method should take a `client.ConfigProvider` not create its own session. Similar pattern to the s3manager's Upload/Download clients.